libopae/api-shell.c

```diff
@@ -269,18 +292,33 @@ fpga_result fpgaGetProperties(fpga_token token, fpga_properties *prop)
 
 		// If the output properties has a parent token set,
 		// then it will be a raw token. We need to wrap it.
-		if (fpgaPropertiesGetParent(*prop, &parent) == FPGA_OK) {
-			opae_wrapped_token *wrapped_parent =
-				opae_allocate_wrapped_token(
-					parent, wrapped_token->adapter_table);
 
-			if (!wrapped_parent) {
-				OPAE_ERR("malloc failed");
-				return FPGA_NO_MEMORY;
-			}
+		p = opae_validate_and_lock_properties(*prop);
 
-			return fpgaPropertiesSetParent(*prop, wrapped_parent);
+		ASSERT_NOT_NULL(p);
+
+		if (FIELD_VALID(p, FPGA_PROPERTY_PARENT)) {
+			fpga_token parent = NULL;
+
+			res = wrapped_token->adapter_table->fpgaCloneToken(
+				p->parent, &parent);
+
+			if (res == FPGA_OK) {
+				opae_wrapped_token *wrapped_parent =
+					opae_allocate_wrapped_token(
+						parent, wrapped_token->adapter_table);
+
+				if (wrapped_parent) {
+					p->parent = wrapped_parent;
+				} else {
+					OPAE_ERR("malloc failed");
+					res = FPGA_NO_MEMORY;
+				}
+
+			}
```

Comment:
We should probably refactor this routine into one function and avoid code duplication

Reply:
Let's wait and see what pans out on the clone discussion.